File: python/mergeSort.py
from random import randint


def createArray(size=10, max=50):
    return [randint(0, max) for _ in range(size)]


def merge(a, b):
    aIdx, bIdx = 0, 0
    merged = []

    while aIdx < len(a) and bIdx < len(b):
        if a[aIdx] < b[bIdx]:
            merged.append(a[aIdx])
            aIdx += 1
        else:
            merged.append(b[bIdx])
            bIdx += 1

    if aIdx == len(a):
        merged.extend(b[bIdx:])
    else:
        merged.extend(a[aIdx])

    return merged


def mergeSort(a):
    # print(a)
    if len(a) <= 1:
        return a

    midpoint = int(len(a)/2)

    left, right = mergeSort(a[:midpoint]), mergeSort(a[midpoint:])

    return merge(left, right)


arr = createArray()
# print(arr)
a = [1, 2, 6]
b = [3, 5, 7]
# merged = merge(a, b)
# print(a[:len(a)//2])
sorted = mergeSort(arr)
print(sorted)

File: ArraySort/js/mergeSort.js
// O(n log n)
/*

Strategy 

1) Base case
2) Split array in half
3) make recusrive call

*/

const mergeSort = (arr) => {
  // Base case
  if (arr.length <= 1) {
    return arr;
  }
  // Split in half
  const midPoint = Math.floor(arr.length / 2);
  const left = arr.slice(0, midPoint);
  const right = arr.slice(midPoint);
  // recursively call left and right
  return merge(mergeSort(left), mergeSort(right));
};

const merge = (left, right) => {
  const merged = [];

  while (left.length && right.length) {
    if (left[0] < right[0]) {
      // Add items to merged and keep removing elements from left
      merged.push(left.shift());
    } else {
      // Add items to merged and keep removing elements from right
      merged.push(right.shift());
    }
  }
  return [...merged, ...left.slice(), ...right.slice()];
};

(function test() {
  arr = [45, 2, 15, 23, 11, 4, 1];
  console.log(mergeSort(arr));
})();
